source/unix/openssl_platform_init.c

```diff
@@ -493,38 +493,76 @@ static enum aws_libcrypto_version s_resolve_libcrypto_symbols(enum aws_libcrypto
-    const char *libcrypto_102 = "libcrypto.so.1.0.0";
-    const char *libcrypto_111 = "libcrypto.so.1.1";
+static enum aws_libcrypto_version s_resolve_libcrypto_compile_version(void) {
+#ifdef OPENSSL_IS_OPENSSL
+    /*
+     * Currently, this only checks for 1.0.2 vs 1.1. As a future optimization, we can also add a branch for OpenSSL 3.0.
+     * OpenSSL 3.0 is compatible with OpenSSL 1.1, so it works currently.
+     */
+    if (OPENSSL_VERSION_NUMBER < 0x10100000L) {
+        return AWS_LIBCRYPTO_1_0_2;
+    } else {
+        return AWS_LIBCRYPTO_1_1_1;
+    }
+#endif
+    return AWS_LIBCRYPTO_NONE;
+}
 
-    AWS_LOGF_DEBUG(AWS_LS_CAL_LIBCRYPTO_RESOLVE, "loading libcrypto 1.0.2");
-    void *module = dlopen(libcrypto_102, RTLD_NOW);
+static char *s_resolve_libcrypto_path_from_enum(enum aws_libcrypto_version version) {
+    switch (version) {
+        case AWS_LIBCRYPTO_1_0_2:
+            return "libcrypto.so.1.0.0";
+        case AWS_LIBCRYPTO_1_1_1:
+            return "libcrypto.so.1.1";
+        default:
+            return "libcrypto.so";
+    }
+}
+
+static enum aws_libcrypto_version s_resolve_libcrypto_lib_impl(enum aws_libcrypto_version version) {
+    const char *libcrypto_version = s_resolve_libcrypto_path_from_enum(version);
+
+    AWS_LOGF_DEBUG(AWS_LS_CAL_LIBCRYPTO_RESOLVE, "loading %s", libcrypto_version);
+    void *module = dlopen(libcrypto_version, RTLD_NOW);
     if (module) {
-        AWS_LOGF_DEBUG(AWS_LS_CAL_LIBCRYPTO_RESOLVE, "resolving against libcrypto 1.0.2");
-        enum aws_libcrypto_version result = s_resolve_libcrypto_symbols(AWS_LIBCRYPTO_1_0_2, module);
-        if (result == AWS_LIBCRYPTO_1_0_2) {
+        AWS_LOGF_DEBUG(AWS_LS_CAL_LIBCRYPTO_RESOLVE, "resolving against %s", libcrypto_version);
+        enum aws_libcrypto_version result = s_resolve_libcrypto_symbols(version, module);
+        if (result == version) {
             return result;
         }
         dlclose(module);
     } else {
-        AWS_LOGF_DEBUG(AWS_LS_CAL_LIBCRYPTO_RESOLVE, "libcrypto 1.0.2 not found");
+        AWS_LOGF_DEBUG(AWS_LS_CAL_LIBCRYPTO_RESOLVE, "%s not found", libcrypto_version);
     }
 
-    AWS_LOGF_DEBUG(AWS_LS_CAL_LIBCRYPTO_RESOLVE, "loading libcrypto 1.1.1");
-    module = dlopen(libcrypto_111, RTLD_NOW);
-    if (module) {
-        AWS_LOGF_DEBUG(AWS_LS_CAL_LIBCRYPTO_RESOLVE, "resolving against libcrypto 1.1.1");
-        enum aws_libcrypto_version result = s_resolve_libcrypto_symbols(AWS_LIBCRYPTO_1_1_1, module);
+    return AWS_LIBCRYPTO_NONE;
+}
+
+static enum aws_libcrypto_version s_resolve_libcrypto_lib(void) {
+    const enum aws_libcrypto_version compiled_version = s_resolve_libcrypto_compile_version();
+    if (compiled_version != AWS_LIBCRYPTO_NONE) {
+        enum aws_libcrypto_version result = s_resolve_libcrypto_lib_impl(compiled_version);
+        if (result == compiled_version) {
+            return result;
+        }
+    }
+
+    if (compiled_version != AWS_LIBCRYPTO_1_1_1) {
```

Comment:
im a bit confused about logic here.
if its not openssl then we resolve compile version to none and otherwise we resolved to one of openssl versions.

so in case of openssl 1.1.1 
compile version is AWS_LIBCRYPTO_1_1_1
so it skips this if and goes to the next if where it loads 1.0.2?

Reply:
If we resolve to compiled version 1.0.2 or 1.1, we try to load it first in the previous if. This if basically just skips trying to reload it if the compiled version was same.